qiskit/synthesis/discrete_basis/solovay_kitaev.py

```diff
@@ -0,0 +1,257 @@
+
+"""Synthesize a single qubit gate to a discrete basis set."""
+
+from __future__ import annotations
+
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit, Gate
+from qiskit.converters import circuit_to_dag
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.dagcircuit.dagcircuit import DAGCircuit
+from qiskit.transpiler.exceptions import TranspilerError
+
+from .gate_sequence import GateSequence
+from .commutator_decompose import commutator_decompose
+from .generate_basis_approximations import generate_basic_approximations, _1q_gates, _1q_inverses
+
+
+class SolovayKitaevDecomposition:
+    """The Solovay Kitaev discrete decomposition algorithm.
+
+    This class is called recursively by the transpiler pass, which is why it is separeted.
+    See :mod:`qiskit.synthesis.discrete_basis` for more information.
+    """
+
+    def __init__(
+        self, basic_approximations: str | dict[str, np.ndarray] | list[GateSequence] | None = None
+    ) -> None:
+        """
+
+        .. note::
+
+            If ``basis_gates`` and ``depth`` are not passed, the basic approximations can be
+            generated with the ``generate_basic_approximations`` method and loaded into the
+            class via ``load_basic_approximations``. Since in practice, large basic approximations
+            are required we suggest to generate a sufficiently large set once and always load
+            the approximations afterwards.
+
+        Args:
+            basis_gates: The basis gates used in the basic approximations.
+            depth: The maximum depth of the basic approximations.
+        """
+        if basic_approximations is None:
+            # generate a default basic approximation
+            basic_approximations = generate_basic_approximations(
+                basis_gates=["h", "t", "tdg"], depth=10
+            )
+
+        self.basic_approximations = self.load_basic_approximations(basic_approximations)
+
+    def load_basic_approximations(self, data: dict | str) -> list[GateSequence]:
+        """Load basic approximations.
+
+        Args:
+            data: If a string, specifies the path to the file from where to load the data.
+                If a dictionary, directly specifies the decompositions as ``{gates: matrix}``.
+                There ``gates`` are the names of the gates producing the SO(3) matrix ``matrix``,
+                e.g. ``{"h t": np.array([[0, 0.7071, -0.7071], [0, -0.7071, -0.7071], [-1, 0, 0]]}``.
+
+        Returns:
+            A list of basic approximations as type ``GateSequence``.
+
+        Raises:
+            ValueError: If the number of gate combinations and associated matrices does not match.
+        """
+        # is already a list of GateSequences
+        if isinstance(data, list):
+            return data
+
+        # if a file, load the dictionary
+        if isinstance(data, str):
+            data = np.load(data, allow_pickle=True)
```

Comment:
Is there a reason to keep the save functionality around since the computation is much faster now?

Reply:
Yeah, it's likely still slow for large basis sets/depths or you'd quickly like to restart a calculation 🙂